src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -3409,6 +3409,8 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
                     else if (treeInfo.isPartialFunctionMissingParamType(tree)) {
                       if (argPtAlts.exists(ts => isPartialFunctionType(ts._1))) partialFunctionProto(argPtAlts)
                       else functionProto(argPtAlts)
+                    } else if (!treeInfo.isFunctionWithParamType(tree) && argPtAlts.forall { case (t, _) => isFunctionType(t)}) {
+                      functionProto(argPtAlts)
```

Comment:
Could you add a comment which case is handled here? And while we're at it, it would be great to have a comment explaining why / when we use a `partialFunctionProto` above.